rust/kernel/thread.rs

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::bindings;
+use crate::c_types;
+
+use alloc::boxed::Box;
+use core::ops::FnOnce;
+
+#[no_mangle]
+unsafe extern "C" fn rust_thread_func(data: *mut c_types::c_void) -> c_types::c_int {
+    Box::from_raw(data as *mut Box<dyn FnOnce()>)();
+    0
+}
+
+pub struct Thread {
+    task: *mut bindings::task_struct,
+}
+
+impl Thread {
+    pub fn new<F>(f: F, name: &str) -> Self
+    where
+        F: FnOnce(),
+        F: Send + 'static,
+    {
+        let bf: Box<dyn FnOnce() + 'static> = Box::new(f);
```

Comment:
Please use a `Box::try_new`.

Actually, can we do this without allocating memory? Something like: allocate from the stack, create the thread, wait for thread init to complete. Thread init move data from the creator stack to its own, informs the creator that it's done with init, then runs the function.

@ojeda -- can we use the clippy to ban `Box::new` and `Arc::new`?

Reply:
I will take a look. We will likely want to have our own clippy set of checks, or at the very least put it in the tidy script.

Medium-term, we will likely want to have our own `alloc` crate.

Reply:
@wedsonaf Got it, I will use `Box::try_new`.

And I will look into whether I could avoid allocating memory, but please note that kernel's `kthread_create_on_node` will simply add the thread creation information to a global list, which later on is fetched by `kthreadd_task` (normally it's the `init` process, until then the thread creation information may not be used. I can combine `kthread_create_on_node` and `wake_up_process` together to make `Thread::new` blocked until the new thread gets initialized, but a) it will make `Thread::new` only work in sleeplable context and b) there are places users want to do creation and wakeup separately.

Thoughts?

Reply:
Hmm.. just check on `kthread_create_on_node`, it will wait until the new thread gets initialized. So it's not a blocker for avoiding box.

Reply:
We could provide both constructors. In my view, it is fine to have helper APIs as long as they are properly documented and they cover a common use case.

Reply:
Our existing allocator is essentially `kmalloc(GFP_KERNEL)`, so we can't allocate from atomic context at the moment.

Leaving the above aside, I agree that we should make it work in atomic context.

I'm not well-versed on how rust allocates closures, but could we have two (or more) constructors, for example, a more restrictive one where a static function plus one argument would be passed (this would involve no allocation from us, and would always work on any context), and a more general one that captures arbitrary context but requires allocation (which for now would only work in sleepable context, but that we could fix in the future)?

The reason for this is mostly to try to provide a zero-cost version. Otherwise, people may argue that rust has non-optional extra costs that C doesn't have. By having two versions, we could argue that we have a C-equivalent version (in terms of performance), but if you want additional ease of use, there is a small performance penalty to pay.

Reply:
Oops, I had a stale version of this comment.

Yes, looking at the implementation it does already do the stack-allocated dance.

Reply:
I was missing something in my previous comment: `kthread_create_on_node` will call `kmalloc(GFP_KERNEL)` to allocate `kthread_create_info`, that makes it not callable in atomic context. Plus, it will **wait** for `kthreadd_task` to complete the creation, which also suggests that `kthread_create_on_node` should only be used in sleepable contexts.

Reply:
This is one more reason why need to take this chance to raise the bar on doc quality in the kernel: documenting under which contexts *all* APIs can be used would be great for both callers and people changing the implementation of a given function. :)

Reply:
The problem of documentation is that it's challenging to sync with code ;-)

So I wonder whether there is something we can do in Rust to annotate a function or code block that may (or maynot sleep), something like:

```
#[maybe(sleep)]
pub fn func1(...) -> ... {
    ...
}

pub fn func2(...) -> ... {
    maybe_sleep!{
    <code here>
    }
}

#[maynot(sleep)] // we can make every function defaultly not sleepable
pub fn func3(...) -> .... {
     
}
```

so that if some one code like:

```
#[maynot(sleep)] // we can make every function defaultly not sleepable
pub fn func3(...) -> .... {
     func1();
}
```

or 

```
#[maynot(sleep)] // we can make every function defaultly not sleepable
pub fn func3(...) -> .... {
     func2();
}
```

compiler will yell.

Thoughts?

Reply:
> The problem of documentation is that it's challenging to sync with code ;-)

Indeed! However, that depends mostly on several factors:
  - How far the docs are from the source, e.g. putting things in `Doc/` is a very easy way indeed to make them out of date, as you say! However, if they are in the function itself, it is *way* harder to miss.
  - How prevalent the style and enforcement is. If documentation is "best-effort" or very inconsistent (like in many of the C APIs we have right now in the kernel), then, as you say, it is likely to go out of sync. However, if we do it consistently, everywhere, and we "call out" any modifications that mistakenly do so, then any developer modifying the constraint *will* know they need to review the docs.
  - How "beautiful" docs are. You may laugh ;-),  but having nice, inline docs that everyone reads (like the Rust `std` ones) makes it *much* more likely that others will contribute and spot bugs on them. You can see this effect happening on the Rust `std` itself!

That is why it is my intention to raise the bar on documentation (and why I am a bit annoying in the PRs ;-).

> So I wonder whether there is something we can do in Rust to annotate a function or code block that may (or maynot sleep), something like:

This is actually something we discussed in the past (see #3)! It could be marked potentially in several ways: colored `unsafe`, custom attributes as you suggest, etc.; and could be enforced differently (e.g. clippy pass for custom attributes; or an entire language feature for colored `unsafe`).

It is something I definitely want to attempt in one way or another medium/long-term, but I think for the moment we should, nevertheless, do it in the docs. It is already useful, and will make the transition later to a "proper feature" if we are successful mainlining this! :-)

Reply:
@wedsonaf I still use box because I need to keep the thread closure living longer than the `Thread::new`. May need some more investigation on how to avoid using box.
@ojeda Add doc for double boxing and others.